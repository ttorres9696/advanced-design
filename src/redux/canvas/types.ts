import { Vector2d } from 'konva/types/types';

import { ShapeCoords } from '../../shared/components/ShapeElement/types';

export const SELECT_SHAPE = 'SELECT_SHAPE';
export const DESELECT_SHAPE = 'DESELECT_SHAPE';
export const INCREASE_ZOOM = 'INCREASE_ZOOM';
export const DECREASE_ZOOM = 'DECREASE_ZOOM';
export const NORMALIZE_ZOOM = 'NORMALIZE_ZOOM';
export const SET_CANVAS_ORIGIN = 'SET_CANVAS_ORIGIN';
export const INCREMENT_TOTAL_SHAPES = 'INCREMENT_TOTAL_SHAPES';
export const DECREMENT_TOTAL_SHAPES = 'DECREMENT_TOTAL_SHAPES';
export const SET_NEW_DRAGGABLE_POINT_MODE = 'SET_NEW_DRAGGABLE_POINT_MODE';
export const SET_MOUSE_POSITION = 'SET_MOUSE_POSITION';
export const SET_DELETE_DRAGGABLE_POINT_MODE = 'SET_DELETE_DRAGGABLE_POINT_MODE';
export const TOGGLE_MAGNET_MODE = 'TOGGLE_MAGNET_MODE';
export const SET_DRAGGING = 'SET_DRAGGING';
export const SET_SELECT_MODE = 'SET_SELECT_MODE';

export const originCoords = 20;
export type SelectionType = 'reset' | 'add';

interface SelectShapeAction {
  type: typeof SELECT_SHAPE;
  payload: {
    solarElementId: string;
    layer: string;
    selectionType: SelectionType;
  };
}

interface DeselectShapeAction {
  type: typeof DESELECT_SHAPE;
  payload: {
    solarElementId: string;
  };
}

interface IncreaseZoomAction {
  type: typeof INCREASE_ZOOM;
}

interface DecreaseZoomAction {
  type: typeof DECREASE_ZOOM;
}

interface NormalizeZoomAction {
  type: typeof NORMALIZE_ZOOM;
}

interface SetCanvasOriginAction {
  type: typeof SET_CANVAS_ORIGIN;
  payload: ShapeCoords;
}

interface IncrementTotalShapesAction {
  type: typeof INCREMENT_TOTAL_SHAPES;
}

interface DecrementTotalShapesAction {
  type: typeof DECREMENT_TOTAL_SHAPES;
  payload: number;
}

interface SetNewDraggablePointModeAction {
  type: typeof SET_NEW_DRAGGABLE_POINT_MODE;
  payload: boolean;
}

interface SetDeleteDraggablePointMode {
  type: typeof SET_DELETE_DRAGGABLE_POINT_MODE;
  payload: boolean;
}

interface SetMousePositionAction {
  type: typeof SET_MOUSE_POSITION;
  payload?: Vector2d;
}

interface ToggleMagnetModeAction {
  type: typeof TOGGLE_MAGNET_MODE;
}

interface SetDragging {
  type: typeof SET_DRAGGING;
  payload: boolean;
}

interface SetSelectMode {
  type: typeof SET_SELECT_MODE;
  payload: boolean;
}

export type CanvasActionTypes =
  | SelectShapeAction
  | DeselectShapeAction
  | IncreaseZoomAction
  | DecreaseZoomAction
  | NormalizeZoomAction
  | SetCanvasOriginAction
  | IncrementTotalShapesAction
  | DecrementTotalShapesAction
  | SetNewDraggablePointModeAction
  | SetDeleteDraggablePointMode
  | SetMousePositionAction
  | ToggleMagnetModeAction
  | SetDragging
  | SetSelectMode;
