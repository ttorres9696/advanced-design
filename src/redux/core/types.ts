import { ModuleSpec } from '../../shared/models/SolarDesign.interface';
import { DeleteDialogParams, WindowDimension } from './state';

export const FETCHED_MODULE_SPECS = 'FETCHED_MODULE_SPECS';
export const SET_DELETE_DIALOG_PARAMS = 'SET_DELETE_DIALOG_PARAMS';
export const SET_WINDOW_DIMENSION = 'SET_WINDOW_DIMENSION';
export const SET_PRE_LOADING = 'SET_PRE_LOADING';
export const CANCEL_PRE_LOADING = 'CANCEL_PRE_LOADING';
export const DECREASE_PRODUCTION_VALUE = 'DECREASE_PRODUCTION_VALUE';
export const INCREASE_PRODUCTION_VALUE = 'INCREASE_PRODUCTION_VALUE';
export const RESET_PRODUCTION_VALUE = 'RESET_PRODUCTION_VALUE';

interface FetchedModuleSpecsAction {
  type: typeof FETCHED_MODULE_SPECS;
  payload: ModuleSpec[];
}

interface SetDeleteDialogParamsAction {
  type: typeof SET_DELETE_DIALOG_PARAMS;
  payload: DeleteDialogParams;
}

interface SetWindowDimensionAction {
  type: typeof SET_WINDOW_DIMENSION;
  payload: WindowDimension;
}

interface SetPreLoadingAction {
  type: typeof SET_PRE_LOADING;
  payload: boolean;
}

interface CancelPreLoading {
  type: typeof CANCEL_PRE_LOADING;
}

interface DecreaseProductionValue {
  type: typeof DECREASE_PRODUCTION_VALUE;
  payload: number;
}

interface IncreaseProductionValue {
  type: typeof INCREASE_PRODUCTION_VALUE;
  payload: number;
}

interface ResetProductionValue {
  type: typeof RESET_PRODUCTION_VALUE;
}

export type CoreActionTypes =
  | FetchedModuleSpecsAction
  | SetDeleteDialogParamsAction
  | SetWindowDimensionAction
  | SetPreLoadingAction
  | CancelPreLoading
  | DecreaseProductionValue
  | IncreaseProductionValue
  | ResetProductionValue;
