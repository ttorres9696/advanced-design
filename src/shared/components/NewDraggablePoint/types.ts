import { Vector2d } from 'konva/types/types';

export interface NewDraggablePointProps {
  newDraggablePointMode: boolean;
  mousePosition?: Vector2d;
}
