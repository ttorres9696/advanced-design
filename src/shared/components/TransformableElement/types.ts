import Konva from 'konva';
import { Node } from 'konva/types/Node';
import { ReactElement } from 'react';

import { SelectionType } from '../../../redux/canvas/types';
import { ShapeElementType } from '../ShapeElement/types';

export interface UseTansfromableElementProps {
  id: string;
  layer: string;
  onStartChange: (id: string, layer: string) => void;
  onChange: (id: string, layer: string, params: TransformableElementChangeParams) => void;
  onFinishChange: (id: string, layer: string) => void;
  onRotate: (id: string, layer: string, pointsRotationInRadians: number[], points: number[][]) => void;
  setDragging: (dragging: boolean) => void;
  version: number;
  deleteDraggablePointMode: boolean;
  onMoveStart: (layer: string, node: Node) => void;
  onMove: (id: string, layer: string, node: Node, e: Konva.KonvaEventObject<DragEvent>) => void;
  onMoveEnd: (layer: string, node: Node) => void;
  onSelect: (id: string, layer: string, selectionType: SelectionType, node: Node) => void;
  visible: boolean;
  locked: boolean;
  multiple: boolean;
  selectMode: boolean;
}

export interface TransformableElementProps extends UseTansfromableElementProps {
  shapeType: ShapeElementType;
  resizeEnabled?: boolean;
  children?: ReactElement | ReactElement[];
  isSelected: boolean;
  active: boolean;
}

export interface TransformableElementChangeParams {
  x?: number;
  y?: number;
  scaleX?: number;
  scaleY?: number;
  width?: number;
  height?: number;
  radius?: number;
  points?: number[][];
  pointsRotationInRadians?: number[];
}

export interface TransformerDimensions {
  width: number;
  height: number;
}
