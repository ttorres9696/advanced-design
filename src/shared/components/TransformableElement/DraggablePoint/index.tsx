import RED from '@material-ui/core/colors/red';
import { Circle as CircleType } from 'konva/types/shapes/Circle';
import { Stage } from 'konva/types/Stage';
import React, { Fragment, useCallback, useRef, useState } from 'react';
import { Circle } from 'react-konva';

import { DraggablePointProps } from './types';

const DraggablePoint: React.FC<DraggablePointProps> = ({
  x,
  y,
  position,
  onStart,
  onChange,
  onEnd,
  onClick,
  deleteDraggablePointMode,
  invisible,
}) => {
  const shapeRef = useRef<any>();

  const [mouseOver, setMouseOver] = useState<boolean>(false);

  const onDragStart = useCallback((e: any) => onStart(), [onStart]);

  const onDragMove = useCallback(
    (e: any) => {
      const node = shapeRef.current;
      onChange(node.x(), node.y(), position);
    },
    [onChange, shapeRef, position],
  );

  const onDragEnd = useCallback(
    (e: any) => {
      const node = shapeRef.current;
      onEnd(node.x(), node.y(), position);
    },
    [onEnd, shapeRef, position],
  );

  const onClickPoint = useCallback((e: any) => onClick(position), [onClick, position]);

  const onMouseOver = useCallback(
    (e: any) => {
      if (deleteDraggablePointMode) {
        setMouseOver(true);
      }
    },
    [deleteDraggablePointMode, setMouseOver],
  );

  const onMouseOut = useCallback((e: any) => setMouseOver(false), [setMouseOver]);

  const onMouseEnter = useCallback(() => {
    if (shapeRef && shapeRef.current) {
      const node = shapeRef.current as CircleType;
      const stage: Stage | null = node.getStage();

      if (stage) {
        stage.container().style.cursor = deleteDraggablePointMode ? 'pointer' : 'move';
      }
    }
  }, [shapeRef, deleteDraggablePointMode]);

  const onMouseLeave = useCallback(() => {
    if (shapeRef && shapeRef.current) {
      const node = shapeRef.current as CircleType;
      const stage: Stage | null = node.getStage();

      if (stage) {
        stage.container().style.cursor = 'default';
      }
    }
  }, [shapeRef]);

  return (
    <Fragment>
      <Circle
        ref={shapeRef}
        x={x}
        y={y}
        radius={deleteDraggablePointMode && mouseOver ? 7 : 5}
        strokeWidth={0.5}
        tension={0}
        stroke={deleteDraggablePointMode && mouseOver ? RED[500] : '#2db1ff'}
        fill={deleteDraggablePointMode && mouseOver ? RED[50] : '#fff'}
        opacity={1}
        draggable={!deleteDraggablePointMode}
        onDragStart={onDragStart}
        onDragMove={onDragMove}
        onDragEnd={onDragEnd}
        onClick={onClickPoint}
        onMouseOver={onMouseOver}
        onMouseOut={onMouseOut}
        onMouseEnter={onMouseEnter}
        onMouseLeave={onMouseLeave}
        visible={!invisible}
      />
    </Fragment>
  );
};

export default React.memo(DraggablePoint);
