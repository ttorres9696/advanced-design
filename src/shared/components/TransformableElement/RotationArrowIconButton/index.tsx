import { Image as ImageType } from 'konva/types/shapes/Image';
import { Stage } from 'konva/types/Stage';
import React, { Fragment, useCallback, useEffect, useRef, useState } from 'react';
import { Image } from 'react-konva';

import { redoBlack } from '../../../assets/redoBlack';
import { undoBlack } from '../../../assets/undoBlack';
import { RotationArrowIconButtonProps, RotationIconSize } from './types';

const RotationArrowIconButton: React.FC<RotationArrowIconButtonProps> = ({
  x,
  y,
  position,
  onClick,
  type,
  rotation,
  invisible,
}) => {
  const shapeRef = useRef<any>();

  const [mouseOver, setMouseOver] = useState<boolean>(false);
  const [icon, setIcon] = useState<HTMLImageElement>();
  const [loadedIcon, setLoadedIcon] = useState<boolean>(false);

  const onClickIcon = useCallback((e: any) => onClick(position), [onClick, position]);

  const onMouseOver = useCallback(
    (e: any) => {
      setMouseOver(true);
    },
    [setMouseOver],
  );

  const onMouseOut = useCallback((e: any) => setMouseOver(false), [setMouseOver]);

  const onMouseEnter = useCallback(() => {
    if (shapeRef && shapeRef.current) {
      const node = shapeRef.current as ImageType;
      const stage: Stage | null = node.getStage();

      if (stage) {
        stage.container().style.cursor = 'pointer';
      }
    }
  }, [shapeRef]);

  const onMouseLeave = useCallback(() => {
    if (shapeRef && shapeRef.current) {
      const node = shapeRef.current as ImageType;
      const stage: Stage | null = node.getStage();

      if (stage) {
        stage.container().style.cursor = 'default';
      }
    }
  }, [shapeRef]);

  const handleImageLoad = useCallback(() => {
    setLoadedIcon(true);
  }, [setLoadedIcon]);

  useEffect(() => {
    if (!icon) {
      const image = new window.Image();
      image.src = type === 'left' ? undoBlack : redoBlack;
      image.addEventListener('load', handleImageLoad);

      setIcon(image);
    }
  }, [icon, setIcon, redoBlack, undoBlack, handleImageLoad]);

  return (
    <Fragment>
      {loadedIcon ? (
        <Image
          ref={shapeRef}
          image={icon}
          x={x}
          y={y}
          width={RotationIconSize}
          height={RotationIconSize}
          tension={0}
          opacity={mouseOver ? 1 : 0.5}
          draggable={false}
          rotation={rotation}
          onClick={onClickIcon}
          onMouseOver={onMouseOver}
          onMouseOut={onMouseOut}
          onMouseEnter={onMouseEnter}
          onMouseLeave={onMouseLeave}
          visible={!invisible}
        />
      ) : null}
    </Fragment>
  );
};

export default React.memo(RotationArrowIconButton);
