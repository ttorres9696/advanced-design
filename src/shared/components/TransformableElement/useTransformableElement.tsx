import Konva from 'konva';
import { Node } from 'konva/types/Node';
import { Line } from 'konva/types/shapes/Line';
import { Stage } from 'konva/types/Stage';
import { useCallback, useEffect, useRef, useState } from 'react';

import { getCurrentModulePointsRotation } from '../../helpers/modules.helper';
import { getRotatedPoints } from '../../helpers/polygons.helper';
import { convertVector2dToPoint } from '../../helpers/vectors.helper';
import { DraggablePointParams } from './DraggablePoint/types';
import { UseTansfromableElementProps } from './types';

const useTransformableElement = ({
  id,
  layer,
  onStartChange,
  onChange,
  onFinishChange,
  onRotate,
  setDragging,
  version,
  deleteDraggablePointMode,
  onMoveStart,
  onMove,
  onMoveEnd,
  onSelect,
  visible,
  locked,
  selectMode,
}: UseTansfromableElementProps) => {
  const shapeRef = useRef<any>();

  const [inDragMode, setInDragMode] = useState<boolean>(false);
  const [draggablePoints, setDraggablePoints] = useState<DraggablePointParams[]>([]);
  const [previousVersion, setPreviousVersion] = useState<number>(0);

  const defineDraggablePoints = useCallback(
    (shapePoints: number[]) => {
      setDraggablePoints(
        shapePoints.reduce((draggablePoints: DraggablePointParams[], coord: number, index: number) => {
          if (index % 2 === 0) {
            draggablePoints.push({
              x: coord,
              y: shapePoints[index + 1],
              position: index / 2,
            });
          }

          return draggablePoints;
        }, []),
      );
    },
    [setDraggablePoints],
  );

  const onDragStart = useCallback(
    (e: any) => {
      if (typeof e === 'object' && e.type && e.type === 'dragstart') {
        const node = shapeRef.current;

        onMoveStart(layer, node);
        setInDragMode(true);
        setDragging(true);
      }
    },
    [layer, onMoveStart, setInDragMode, setDragging],
  );

  const onDragMove = useCallback(
    (e: any) => {
      if (typeof e === 'object' && e.type && e.type.indexOf('drag') === 0) {
        onMove(id, layer, shapeRef.current, e);
      }
    },
    [id, layer, onMove, shapeRef],
  );

  const onDragEnd = useCallback(
    (e: any) => {
      const node = shapeRef.current;

      onMoveEnd(layer, node);
      setInDragMode(false);
      setDragging(false);
    },
    [layer, shapeRef, setInDragMode, setDragging, onMoveEnd],
  );

  const onPolygonTransformStart = useCallback(() => onStartChange(id, layer), [onStartChange, id, layer]);

  const onPolygonTransformEnd = useCallback(
    (e: any) => {
      const node = shapeRef.current as Line;

      const rotation = node.rotation();
      node.rotate(-1 * rotation);
      node.x(0);
      node.y(0);

      const rotatedPoints = getRotatedPoints(
        rotation,
        draggablePoints.map(draggablePoint => convertVector2dToPoint(draggablePoint)),
      );

      onRotate(id, layer, getCurrentModulePointsRotation(rotatedPoints), rotatedPoints);

      onFinishChange(id, layer);
    },
    [shapeRef, id, layer, draggablePoints, onRotate, onFinishChange],
  );

  const onCircleTransformStart = useCallback(() => onStartChange(id, layer), [onStartChange, id, layer]);

  const onCircleTransformEnd = useCallback(
    (e: any) => {
      const node = shapeRef.current;
      const scaleX = node.scaleX();

      node.scaleX(1);
      node.scaleY(1);
      onChange(id, layer, {
        radius: node.radius() * scaleX,
      });

      onFinishChange(id, layer);
    },
    [onChange, onFinishChange, id, layer, shapeRef],
  );

  const onPolygonChangePointStart = useCallback(() => onStartChange(id, layer), [onStartChange, id, layer]);

  const onPolygonChangePoint = useCallback(
    (x: number, y: number, position: number) => {
      onChange(id, layer, {
        points: draggablePoints.map((draggablePoint: DraggablePointParams, index: number) =>
          index === position ? [x, y] : [draggablePoint.x, draggablePoint.y],
        ),
      });
    },
    [onChange, id, layer, draggablePoints],
  );

  const onPolygonChangePointEnd = useCallback(
    (x: number, y: number, position: number) => {
      setDraggablePoints(
        draggablePoints.map((draggablePoint: DraggablePointParams, index: number) =>
          index === position ? { x, y, position } : draggablePoint,
        ),
      );

      onFinishChange(id, layer);
    },
    [onFinishChange, id, layer, draggablePoints, setDraggablePoints],
  );

  const onPolygonClickPoint = useCallback(
    (position: number) => {
      if (deleteDraggablePointMode && shapeRef && shapeRef.current) {
        onStartChange(id, layer);

        const newDraggablePoints = Array.from(draggablePoints);
        newDraggablePoints.splice(position, 1);

        onChange(id, layer, {
          points: newDraggablePoints.map(draggablePoint => [draggablePoint.x, draggablePoint.y]),
        });

        setDraggablePoints(newDraggablePoints);

        onFinishChange(id, layer);

        const node = shapeRef.current as Node;
        const stage: Stage | null = node.getStage();

        if (stage) {
          stage.container().style.cursor = 'default';
        }
      }
    },
    [
      onChange,
      id,
      layer,
      draggablePoints,
      setDraggablePoints,
      deleteDraggablePointMode,
      onStartChange,
      onFinishChange,
      shapeRef,
    ],
  );

  const onMouseEnter = useCallback(
    (e: any) => {
      const container = e.target.getStage().container();
      container.style.cursor = layer === 'modules' && !selectMode ? 'pointer' : 'default';
    },
    [layer, selectMode],
  );

  const onMouseLeave = useCallback((e: any) => {
    const container = e.target.getStage().container();
    container.style.cursor = 'default';
  }, []);

  const handleOnClickElement = useCallback(
    (e: Konva.KonvaEventObject<MouseEvent>) => {
      if (shapeRef && shapeRef.current && visible && !locked) {
        onDragStart(id);
        onSelect(id, layer, e.evt.ctrlKey || e.evt.metaKey ? 'add' : 'reset', shapeRef.current as Node);
      }
    },
    [id, layer, shapeRef, onDragStart, onSelect, visible, locked],
  );

  useEffect(() => {
    if (
      (!draggablePoints.length || previousVersion !== version) &&
      shapeRef &&
      shapeRef.current &&
      shapeRef.current.points
    ) {
      defineDraggablePoints(shapeRef.current.points());

      setPreviousVersion(version);
    }
  }, [shapeRef, defineDraggablePoints, draggablePoints.length, previousVersion, setPreviousVersion, version]);

  return {
    shapeRef,
    onDragStart,
    onDragMove,
    onDragEnd,
    onCircleTransformStart,
    onCircleTransformEnd,
    onPolygonTransformStart,
    onPolygonTransformEnd,
    onPolygonChangePointStart,
    onPolygonChangePoint,
    onPolygonChangePointEnd,
    onPolygonClickPoint,
    draggablePoints,
    inDragMode,
    handleOnClickElement,
    onMouseEnter,
    onMouseLeave,
  };
};

export default useTransformableElement;
