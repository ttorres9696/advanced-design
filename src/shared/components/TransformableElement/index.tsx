import React, { Children, cloneElement } from 'react';

import DraggablePoint from './DraggablePoint';
import { DraggablePointParams } from './DraggablePoint/types';
import { TransformableElementProps } from './types';
import useTransformableElement from './useTransformableElement';

const OPACITY_SELECTED = 0.7;
const OPACITY_NOT_SELECTED = 0.5;
const OPACITY_NOT_VISIBLE = 0;
const OPACITY_INACTIVE_MODULE = 0.3;
const OPACITY_INACTIVE_MODULE_IN_SELECT_MODE = 0.1;

const TransformableElement: React.FC<TransformableElementProps> = ({
  id,
  layer,
  children,
  isSelected,
  onSelect,
  onStartChange,
  onChange,
  onRotate,
  onFinishChange,
  onMoveStart,
  onMove,
  onMoveEnd,
  visible,
  locked,
  resizeEnabled,
  shapeType,
  version,
  deleteDraggablePointMode,
  setDragging,
  multiple,
  active,
  selectMode,
}) => {
  const {
    shapeRef,
    onDragStart,
    onDragMove,
    onDragEnd,
    onCircleTransformStart,
    onCircleTransformEnd,
    onPolygonTransformStart,
    onPolygonTransformEnd,
    onPolygonChangePointStart,
    onPolygonChangePoint,
    onPolygonChangePointEnd,
    onPolygonClickPoint,
    draggablePoints,
    inDragMode,
    handleOnClickElement,
    onMouseEnter,
    onMouseLeave,
  } = useTransformableElement({
    id,
    layer,
    onStartChange,
    onChange,
    onFinishChange,
    onRotate,
    setDragging,
    version,
    deleteDraggablePointMode,
    onMoveStart,
    onMove,
    onMoveEnd,
    onSelect,
    visible,
    locked,
    multiple,
    selectMode,
  });

  let draggable = visible && !locked;
  let transformable = isSelected && visible && !locked;

  let opacity = isSelected && visible ? OPACITY_SELECTED : !visible ? OPACITY_NOT_VISIBLE : OPACITY_NOT_SELECTED;

  if (layer === 'modules' && !active) {
    draggable = false;
    transformable = false;
    opacity = selectMode ? OPACITY_INACTIVE_MODULE_IN_SELECT_MODE : OPACITY_INACTIVE_MODULE;
  }

  return (
    <>
      {Children.map(children, child =>
        cloneElement(child as React.ReactElement<any>, {
          ref: shapeRef,
          draggable,
          onClick: handleOnClickElement,
          onDragStart,
          onDragMove,
          onDragEnd,
          onCircleTransformStart,
          onCircleTransformEnd,
          onPolygonTransformStart,
          onPolygonTransformEnd,
          onMouseEnter,
          onMouseLeave,
          opacity,
        }),
      )}
      {transformable ? (
        <>
          {resizeEnabled && shapeType !== 'circle' && shapeRef && shapeRef.current
            ? draggablePoints.map((draggablePoint: DraggablePointParams) => (
                <DraggablePoint
                  key={draggablePoint.position}
                  x={draggablePoint.x}
                  y={draggablePoint.y}
                  position={draggablePoint.position}
                  onStart={onPolygonChangePointStart}
                  onChange={onPolygonChangePoint}
                  onEnd={onPolygonChangePointEnd}
                  onClick={onPolygonClickPoint}
                  deleteDraggablePointMode={deleteDraggablePointMode}
                  invisible={inDragMode}
                />
              ))
            : null}
        </>
      ) : null}
    </>
  );
};

export default React.memo(TransformableElement);
