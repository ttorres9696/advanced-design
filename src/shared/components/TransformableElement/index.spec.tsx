import { mount, ReactWrapper } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Node } from 'konva/types/Node';
import React from 'react';
import { Layer, Line, Stage } from 'react-konva';

import TransformableElement from '.';
import { TransformableElementChangeParams, TransformableElementProps } from './types';

let wrapper: ReactWrapper;

beforeEach(() => {
  const props: TransformableElementProps = {
    id: 'transformable-element-id',
    shapeType: 'polygon',
    layer: 'modules',
    version: 1,
    visible: true,
    locked: false,
    onSelect: (id: string) => {},
    isSelected: true,
    onChange: (id: string, layer: string, params: TransformableElementChangeParams) => {},
    onStartChange: (id: string, layer: string) => {},
    onRotate: (id: string, layer: string, pointsRotationInRadians: number[], points: number[][]) => {},
    onMove: (id: string, layer: string, node: Node) => {},
    onMoveStart: (layer: string, node: Node) => {},
    onMoveEnd: (layer: string, node: Node) => {},
    onFinishChange: (id: string, layer: string) => {},
    deleteDraggablePointMode: false,
    multiple: false,
    setDragging: (dragging: boolean) => {},
    active: true,
    selectMode: true,
  };

  wrapper = mount(
    <Stage>
      <Layer>
        <TransformableElement {...props}>
          <Line x={20} y={20} points={[0, 0, 100, 0, 100, 100]} tension={0} closed stroke="black" fill="lightgray" />
        </TransformableElement>
      </Layer>
    </Stage>,
    {},
  );
});

describe('TransformableElement', () => {
  it('renders correctly', () => expect(toJson(wrapper)).toMatchSnapshot());
});
