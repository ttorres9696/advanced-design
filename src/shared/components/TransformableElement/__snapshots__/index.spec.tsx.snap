// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransformableElement renders correctly 1`] = `
<ForwardRef>
  <Stage
    forwardedRef={null}
  >
    <div />
  </Stage>
</ForwardRef>
`;
